kube-controllers/pkg/controllers/node/labels.go

```diff
@@ -58,161 +104,205 @@ func (c *nodeLabelController) RegisterWith(f *utils.DataFeed) {
 	// The value can be nil even if the UpdateType is New or Updated if it is the result of a
 	// failed validation in the syncer, and we want to treat those as deletes.
-	if update.Value != nil {
-		switch update.KVPair.Value.(type) {
-		case *apiv3.Node:
-			n := update.KVPair.Value.(*apiv3.Node)
-			kn, err := getK8sNodeName(*n)
-			if err != nil {
-				logrus.WithError(err).Info("Unable to get corresponding k8s node name, skipping")
-			} else if kn != "" {
-				// Create a mapping from Kubernetes node -> Calico node.
-				logrus.Debugf("Mapping k8s node -> calico node. %s -> %s", kn, n.Name)
-				c.Lock()
-				c.nodemapper[kn] = n.Name
-				c.Unlock()
+	if update.Value == nil {
+		nodeName := update.Key.(model.ResourceKey).Name
+		// Try to perform unmapping based on resource name (calico node name).
+		for kn, cn := range c.k8sNodeMapper {
+			if cn == nodeName {
+				// Remove it from node map.
+				logrus.Debugf("Unmapping k8s node -> calico node. %s -> %s", kn, cn)
+				delete(c.k8sNodeMapper, kn)
+				delete(c.calicoNodeCache, cn)
+				break
 			}
-		default:
-			// Shouldn't have any other kinds show up here.
-			logrus.Warnf("Unexpected kind received over syncer: %s", update.KVPair.Key)
 		}
-	} else {
-		switch update.KVPair.Key.(type) {
-		case model.ResourceKey:
-			switch update.KVPair.Key.(model.ResourceKey).Kind {
-			case apiv3.KindNode:
-				// Try to perform unmapping based on resource name (calico node name).
-				nodeName := update.KVPair.Key.(model.ResourceKey).Name
-				for kn, cn := range c.nodemapper {
-					if cn == nodeName {
-						// Remove it from node map.
-						logrus.Debugf("Unmapping k8s node -> calico node. %s -> %s", kn, cn)
-						c.Lock()
-						delete(c.nodemapper, kn)
-						c.Unlock()
-						break
-					}
-				}
-			default:
-				// Shouldn't have any other kinds show up here.
-				logrus.Warnf("Unexpected kind received over syncer: %s", update.KVPair.Key)
+		return
+	}
+
+	n := update.Value.(*apiv3.Node)
+	if n.ResourceVersion == "" {
+		n.ResourceVersion = update.Revision
+	}
+
+	kn, err := getK8sNodeName(*n)
+	if err != nil {
+		logrus.WithError(err).Info("Unable to get corresponding k8s node name, skipping")
+	} else if kn != "" {
+		// Create a mapping from Kubernetes node -> Calico node.
+		logrus.Debugf("Mapping k8s node -> calico node. %s -> %s", kn, n.Name)
+
+		c.k8sNodeMapper[kn] = n.Name
+
+		node, err := c.nodeLister.Get(kn)
+		if err != nil {
+			logrus.WithError(err).WithField("node", kn).Error("Unable to get node")
+		}
+
+		if _, ok := c.calicoNodeCache[n.Name]; !ok {
+			// If this is new calicoNode trigger sync for corresponding k8s node.
+			// As we only sync labels k8s -> calico node, no need to sync when calico node already exists in our cache
+			c.calicoNodeCache[n.Name] = n
+
+			if node != nil && c.syncStatus == bapi.InSync {
+				// Only trigger the update if the k8s node exists
+				c.k8sNodeUpdate <- node
 			}
+			return
 		}
 
+		c.calicoNodeCache[n.Name] = n
 	}
```

Comment:
Just a suggestion to exit early, in order to avoid nested conditions.

```suggestion
	if err != nil {
		logrus.WithError(err).Info("Unable to get corresponding k8s node name, skipping")
		return
	} 
	if kn == "" {
		logrus.WithError(err).Info("Corresponding k8s node name is empty, skipping")
		return
	} 	

	// Create a mapping from Kubernetes node -> Calico node.
	logrus.Debugf("Mapping k8s node -> calico node. %s -> %s", kn, n.Name)

	c.k8sNodeMapper[kn] = n.Name

	node, err := c.nodeLister.Get(kn)
	if err != nil {
		logrus.WithError(err).WithField("node", kn).Error("Unable to get node")
	}

	_, existsCalicoNode := c.calicoNodeCache[n.Name];
	c.calicoNodeCache[n.Name] = n
	if  !existsCalicoNode {
		// If this is new calicoNode trigger sync for corresponding k8s node.
		// As we only sync labels k8s -> calico node, no need to sync when calico node already exists in our cache
		if node != nil && c.syncStatus == bapi.InSync {
			// Only trigger the update if the k8s node exists
			c.k8sNodeUpdate <- node
		}
	}
```